src/DecksService/Data/DeckInventoryProvider.cs

```diff
@@ -0,0 +1,295 @@
+                        upsertResult = UpsertResult.UpdatedExisting;
+                    }
+                    else
+                    {
+                        deck.CreatedDateTimeUtc = DateTime.UtcNow;
+                        upsertResult = UpsertResult.CreatedNew;
+                    }
+
+                    IDictionary<string, Card> actualCards = await GetCardInventoryAsync(authToken);
+                    VerifyDeck(deck, actualCards, DateTime.UtcNow);
+                    await WriteDecksToDatabaseAsync(documentClient, new Deck[] { deck });
+                    return new DeckOperationResult<UpsertResult>(upsertResult, deck);
+                }
+            }
+            catch (Exception ex)
+            {
+                throw;
+            }
+        }
+
+        private DocumentClient GetCosmosDbDocumentClient()
+        {
+            var endpointUri = new Uri(this.configuration["DecksDB_Endpoint"]);
+            string authKey = this.configuration["DecksDB_AuthKey"];
+            var documentClient = new DocumentClient(endpointUri, authKey);
+            return documentClient;
+        }
+
+        private async Task<Deck[]> VerifyDecksIfNecessaryAsync(DocumentClient documentClient, Deck[] decks, string authToken)
+        {
+            if (decks == null)
+            {
+                return null;
+            }
+
+            TimeSpan refreshDataAfter = TimeSpan.FromHours(2);
+            DateTime utcNow = DateTime.UtcNow;
+            if (decks.Any(x => utcNow - x.LastVerifiedDateTimeUtc > refreshDataAfter))
+            {
+                IDictionary<string, Card> actualCards = await GetCardInventoryAsync(authToken);
+                VerifyDecks(decks, actualCards, DateTime.UtcNow);
+                await WriteDecksToDatabaseAsync(documentClient, decks);
+            }
+
+            return decks;
+        }
+
+        private Deck[] ReadDecksFromDatabase(DocumentClient documentClient, string userId)
+        {
+            Uri collectionUri = GetDocumentCollectionUri();
+            try
+            {
+                var sqlQuerySpec = new SqlQuerySpec("select * from Items i where i.userId = @userIdParameter");
+                sqlQuerySpec.Parameters.Add(new SqlParameter("@userIdParameter", userId));
+                var feedOptions = new FeedOptions { JsonSerializerSettings = GetJsonSerializerSettings() };
+
+                Deck[] decks = documentClient.CreateDocumentQuery<Deck>(collectionUri, sqlQuerySpec, feedOptions).ToArray();
+                return decks;
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
+            }
+            return null;
+        }
+
+        private async Task<GetDeckResult> ReadDeckFromDatabaseAsync(DocumentClient documentClient, string deckId)
+        {
+            string databaseId = this.configuration["DecksDB_DatabaseName"];
+            string collectionId = this.configuration["DecksDB_Collection"];
```

Comment:
Maybe replace `"DecksDb_DatabaseName"` and such to a constant as they're used a couple of times?